drivers/flash/flash_shell.c

```diff
@@ -255,6 +254,246 @@ static int cmd_test(const struct shell *shell, size_t argc, char *argv[])
+			    CONFIG_FLASH_SHELL_BUFFER_SIZE);
+		return -EINVAL;
+	}
+
+	result = 0;
+
+	while (repeat--) {
+		start_tick = k_uptime_get();
+		result = flash_read(flash_dev, addr, test_arr, size);
+		loop_tick = k_uptime_delta(&start_tick);
+
+		if (result) {
+			shell_error(sh, "Read failed: %d", result);
+			break;
+		}
+
+		++loops;
+		total_tick += loop_tick;
+		shell_print(sh, "Loop #%u done in %llu ticks.", loops, loop_tick);
+	}
+
+	if (result == 0) {
+		speed_output(sh, total_tick, (double)loops, (double)size);
+	}
+
+	return result;
+}
+
+static int cmd_write_test(const struct shell *sh, size_t argc, char *argv[])
+{
+	const struct device *flash_dev;
+	uint32_t repeat;
+	int result;
+	uint32_t addr;
+	uint32_t size;
+	uint64_t start_tick;
+	uint64_t loop_tick;
+	uint64_t total_tick = 0;
+	uint32_t loops = 0;
+
+	result = parse_helper(sh, &argc, &argv, &flash_dev, &addr);
+	if (result) {
+		return result;
+	}
+
+	size = strtoul(argv[2], NULL, 16);
+	repeat = strtoul(argv[3], NULL, 16);
+	if (size > CONFIG_FLASH_SHELL_BUFFER_SIZE) {
+		shell_error(sh, "<size> must be at most 0x%x.",
+			    CONFIG_FLASH_SHELL_BUFFER_SIZE);
+		return -EINVAL;
+	}
+
+	for (uint32_t i = 0; i < size; i++) {
+		test_arr[i] = (uint8_t)i;
+	}
+
+	result = 0;
+
+	while (repeat--) {
+		start_tick = k_uptime_get();
+		result = flash_write(flash_dev, addr, test_arr, size);
+		loop_tick = k_uptime_delta(&start_tick);
+
+		if (result) {
+			shell_error(sh, "Write failed: %d", result);
+			break;
+		}
+
+		++loops;
+		total_tick += loop_tick;
+		shell_print(sh, "Loop #%u done in %llu ticks.", loops, loop_tick);
+	}
+
+	if (result == 0) {
+		speed_output(sh, total_tick, (double)loops, (double)size);
+	}
+
+	return result;
+}
+
+static int cmd_erase_test(const struct shell *sh, size_t argc, char *argv[])
+{
+	const struct device *flash_dev;
+	uint32_t repeat;
+	int result;
+	uint32_t addr;
+	uint32_t size;
+	uint64_t start_tick;
+	uint64_t loop_tick;
+	uint64_t total_tick = 0;
+	uint32_t loops = 0;
+
+	result = parse_helper(sh, &argc, &argv, &flash_dev, &addr);
+	if (result) {
+		return result;
+	}
+
+	size = strtoul(argv[2], NULL, 16);
+	repeat = strtoul(argv[3], NULL, 16);
+	if (size > CONFIG_FLASH_SHELL_BUFFER_SIZE) {
+		shell_error(sh, "<size> must be at most 0x%x.",
+			    CONFIG_FLASH_SHELL_BUFFER_SIZE);
+		return -EINVAL;
+	}
+
+	for (uint32_t i = 0; i < size; i++) {
+		test_arr[i] = (uint8_t)i;
+	}
+
+	result = 0;
+
+	while (repeat--) {
+		start_tick = k_uptime_get();
+		result = flash_erase(flash_dev, addr, size);
+		loop_tick = k_uptime_delta(&start_tick);
+
+		if (result) {
+			shell_error(sh, "Erase failed: %d", result);
+			break;
+		}
+
+		++loops;
+		total_tick += loop_tick;
+		shell_print(sh, "Loop #%u done in %llu ticks.", loops, loop_tick);
+	}
+
+	if (result == 0) {
+		speed_output(sh, total_tick, (double)loops, (double)size);
+	}
+
+	return result;
+}
+
+static int cmd_erase_write_test(const struct shell *sh, size_t argc, char *argv[])
```

Comment:
I do not think this one is needed. There is usually some distance between erase and write in code, so this does not profile any valid sequence.

Testing operations by type is fine as it allows to profile them and for example profile how much improvement you get from implementing page vs segment erase in memories, or profile improvements or how much changes to drivers or API cost us in performance.
Also this can be used to get flash API busy, even if ops happen to flash RAM driver, and at the same time check how the other parts of system behave or are impacted.

Reply:
I used this one to check the difference with what the file system one is doing since that is doing erasing and writing, so it gives a like-for-like rough view of the filesystem overhead